statement.py

```diff
@@ -4,43 +4,272 @@
+            return
+
+        invoices = set()
+        for line in self.lines:
+            if (line.invoice
+                    and line.invoice.currency == self.company.currency):
+                invoices.add(line.invoice)
+        invoice_id2amount_to_pay = {}
+        for invoice in invoices:
+            if invoice.type == 'out':
+                sign = -1
+            else:
+                sign = 1
+            invoice_id2amount_to_pay[invoice.id] = sign * invoice.amount_to_pay
+
+        lines = list(self.lines)
+        for line in lines:
+            if (line.invoice
+                    and line.id
+                    and line.invoice.id in invoice_id2amount_to_pay):
+                amount_to_pay = invoice_id2amount_to_pay[line.invoice.id]
+                if (amount_to_pay
+                        and getattr(line, 'amount', None)
+                        and (line.amount >= 0) == (amount_to_pay <= 0)):
+                    if abs(line.amount) > abs(amount_to_pay):
+                        line.amount = amount_to_pay.copy_sign(line.amount)
+                    else:
+                        invoice_id2amount_to_pay[line.invoice.id] = (
+                            line.amount + amount_to_pay)
+                else:
+                    line.invoice = None
+            self.lines = lines
+
+    @classmethod
+    def validate_origin(cls, origins):
+        pool = Pool()
+        Line = pool.get('account.statement.line')
+        Warning = pool.get('res.user.warning')
+
+        paid_cancelled_invoice_lines = []
+        for origin in origins:
+            paid_cancelled_invoice_lines.extend(l for l in origin.lines
+                if l.invoice and l.invoice.state in {'cancelled', 'paid'})
+
+        if paid_cancelled_invoice_lines:
+            warning_key = Warning.format(
+                'statement_paid_cancelled_invoice_lines',
+                paid_cancelled_invoice_lines)
+            if Warning.check(warning_key):
+                raise StatementValidateWarning(warning_key,
+                    gettext('account_statement'
+                        '.msg_statement_invoice_paid_cancelled'))
+            Line.write(paid_cancelled_invoice_lines, {
+                    'related_to': None,
+                    })
+
+    def validate_amount(self):
+        pool = Pool()
+        Lang = pool.get('ir.lang')
+
+        amount = sum(l.amount for l in self.lines)
+        if amount != self.amount:
+            lang = Lang.get()
+            total_amount = lang.currency(
+                self.amount, self.statement.journal.currency)
+            amount = lang.currency(amount, self.statement.journal.currency)
+            raise StatementValidateError(
+                gettext('account_statement_enable_banking.'
+                    'msg_origin_pending_amount',
+                    origin_amount=total_amount,
+                    line_amount=amount))
+
+    @classmethod
+    def validate_statement(cls, statements):
```

Comment:
Why naming this method `validate_statement`? Do we really want to validate the whole statement or simply the origins is enough?

I think it should just be named `validate_origin()` and accept origin objects as parameter.

Also I think it must have a comment indicating that it is basically code copy & pasted from account_statement but adapted to work at origin level.

Reply:
We need to validate the whole statement, so the name for me is correct.

If we work throw the origins, the statement will always be in draft state. So when we reconcile an origin, we check if its statement has all the origins reconciled and if it's true, we close the statement following the core steps.